pkg/common/interceptors.go

```diff
@@ -0,0 +1,165 @@
+	"google.golang.org/grpc/status"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+const (
+	podNameKey          = "csi.storage.k8s.io/pod.name"
+	podNamespaceKey     = "csi.storage.k8s.io/pod.namespace"
+	ControllerPublish   = "ControllerPublishVolume"
+	ControllerUnpublish = "ControllerUnpublishVolume"
+	NodeStage           = "NodeStageVolume"
+	NodeUnstage         = "NodeUnstageVolume"
+	NodePublish         = "NodePublishVolume"
+	NodeUnpublish       = "NodeUnpublishVolume"
+)
+
+var client *kubernetes.Clientset
+
+type reqWithVolumeId interface {
+	GetVolumeId() string
+}
+
+func init() {
+	client = getK8sClient()
+}
+
+func getK8sClient() *kubernetes.Clientset {
+	config, err := clientcmd.BuildConfigFromFlags(options.MasterURL, options.Kubeconfig)
+	if err != nil {
+		log.Errorf("interceptors: initializing kubernetes config failed: %s", err)
+		return nil
+	}
+	client, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		log.Errorf("interceptors: initializing kubernetes clientset failed: %s", err)
+		return nil
+	}
+	return client
+}
+
+func logGRPC(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
+	log.Infof("GRPC call: %s", info.FullMethod)
+	log.Debugf("GRPC request: %s", protosanitizer.StripSecrets(req))
+	resp, err := handler(ctx, req)
+	if err != nil {
+		log.Errorf("GRPC error: %v", err)
+	} else {
+		log.Debugf("GRPC response: %s", protosanitizer.StripSecrets(resp))
+	}
+	return resp, err
+}
+
+func instrumentGRPC(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler, driverType string) (interface{}, error) {
+	if _, ok := req.(reqWithVolumeId); !ok {
+		return handler(ctx, req)
+	}
+
+	volumeId := req.(reqWithVolumeId).GetVolumeId()
+	method := getMethodName(info.FullMethod)
+	switch method {
+	case ControllerPublish, ControllerUnpublish, NodeStage, NodeUnstage, NodeUnpublish:
+		start := time.Now()
+		resp, err := handler(ctx, req)
+		execTime := time.Since(start).Milliseconds()
+		observeExecTime(execTime, method, driverType, volumeId, err)
+		return resp, err
+	case NodePublish:
+		start := time.Now()
+		resp, err := handler(ctx, req)
+		end := time.Now()
+		execTime := end.Sub(start).Milliseconds()
+		observeExecTime(execTime, method, driverType, volumeId, err)
+		if _, ok := req.(*csi.NodePublishVolumeRequest); !ok {
+			return resp, err
+		}
+		request := req.(*csi.NodePublishVolumeRequest)
+		observeVolumeAttachmentTime(end.UnixMilli(), request, driverType, err)
+		return resp, err
+	}
+	return handler(ctx, req)
+}
+
+func getMethodName(fullName string) string {
+	i := strings.LastIndex(fullName, "/")
+	if !strings.Contains(fullName, "/") || i == len(fullName)-1 {
+		return ""
+	}
+	return fullName[i+1:]
+}
+
+func observeExecTime(time int64, method, driverType, volumeId string, err error) {
```

Comment:
```suggestion
func observeExecTime(time time.Duration, method, driverType, volumeId string, err error) {
```
Just pass in the Duration to avoid any ambiguous about unit.